Add ModernBERT to Transformers

warner-benjamin opened the pull request:
This PR will add ModernBERT to Transformers.

warner-benjamin committed:
initial cut of modernbert for transformers

ArthurZucker commented:
cc @Cyrilvallez ! 🤗

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1249 @@
+        return self.out_drop(self.Wo(attn))
+
+
+class ModernBertFlashAttention2(ModernBertAttention):
+    def __init__(self, config: ModernBertConfig, layer_id: Optional[int] = None):
+        super().__init__(config, layer_id)
+        self.config._attn_implementation = "flash_attention_2"
+        logger.warning_once(
+            "The `ModernBertFlashAttention2` class is deprecated in favor of simply modifying the `config._attn_implementation`"
+            "attribute of the `GemmaAttention` class! It will be removed in v4.48"
+        )
+
+
+class ModernBertSdpaAttention(ModernBertAttention):
+    def __init__(self, config: ModernBertConfig, layer_id: Optional[int] = None):
+        super().__init__(config, layer_id)
+        self.config._attn_implementation = "sdpa"
+        logger.warning_once(
+            "The `ModernBertFlashAttention2` class is deprecated in favor of simply modifying the `config._attn_implementation`"
+            "attribute of the `GemmaAttention` class! It will be removed in v4.48"
+        )
```
I don't think these are necessary, they might just be leftover copy-pastes from Gemma.

tomaarsen commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -945,6 +947,7 @@
         ("funnel", "FunnelForSequenceClassification"),
         ("gemma", "GemmaForSequenceClassification"),
         ("gemma2", "Gemma2ForSequenceClassification"),
+        ("modernbert", "ModernBertForSequenceClassification"),
```
This does not exist yet - I'd love to see this as it allows for training cross-encoder/reranker models.

warner-benjamin replied:
The plan is to support all the standard encoder task heads.

tomaarsen commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -1136,6 +1139,7 @@
         ("funnel", "FunnelForTokenClassification"),
         ("gemma", "GemmaForTokenClassification"),
         ("gemma2", "Gemma2ForTokenClassification"),
+        ("modernbert", "ModernBertForTokenClassification"),
```
Also does not exist yet.

tomaarsen reviewed (commented):
Beyond the obvious (sdpa, eager, flex attention, and documentation), I haven't seen anything outrageous or very unexpected in my first scroll-through.
I recognize that this implementation goes a bit beyond our "usual" with unpadding/padding when possible, but I personally don't mind. Beyond this change (and the other obvious upgrades like RoPE), I quite like how this still mirrors the original BERT rather closely.

I'll have to actually start running this to get a better feel, but so far so good.

Also, the SequenceClassification and TokenClassification classes don't exist yet.

tomaarsen commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -203,6 +203,7 @@
                     "GemmaTokenizerFast" if is_tokenizers_available() else None,
                 ),
             ),
+            ("modernbert", (None, "GPTNeoXTokenizerFast" if is_tokenizers_available() else None)),
```
```suggestion
            ("modernbert", (None, "PreTrainedTokenizerFast" if is_tokenizers_available() else None)),
```
`GPTNeoXTokenizerFast` differs from what was used during training.
We could also use`ModernBertTokenizerFast` depending on my previous comment.

warner-benjamin replied:
For the moment, I have it set as `PreTrainedTokenizerFast` pending discussion.

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1249 @@
+        batch_size: Optional[int] = None,
+        seq_len: Optional[int] = None,
+    ) -> Union[Tuple[torch.Tensor, ...], BaseModelOutput]:
+        if batch_size is None and seq_len is None:
```
```suggestion
	return_dict = return_dict if return_dict is not None else self.config.use_return_dict
	
        if batch_size is None and seq_len is None:
```
Required, or the `None` default will always cause the returning of a tuple (which is unexpected).

warner-benjamin replied:
fixed

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1249 @@
+        hidden_states = self.final_norm(hidden_states)
+
+        if repad:
+            hidden_states = self._pad_outputs(hidden_states, indices, batch_size, seq_len)
```
```suggestion
            hidden_states, _ = self._pad_outputs(hidden_states, indices, batch_size, seq_len)
```
This also returns padded `labels`, which we don't use.

warner-benjamin replied:
fixed

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1249 @@
+
+        if self.pad_logits:
+            if self.pad_logits_no_grad:
+                logits = self._pad_outputs_no_grad(logits, indices, batch_size, seq_len)[0]
+            else:
+                logits = self._pad_outputs(logits, indices, batch_size, seq_len)[0]
```
```suggestion
                logits, _ = self._pad_outputs_no_grad(logits, indices, batch_size, seq_len)[0]
            else:
                logits, _ = self._pad_outputs(logits, indices, batch_size, seq_len)[0]
```
This also returns padded `labels`, which we don't use.

warner-benjamin replied:
The [0] index should have taken care of that, but I matched the formatting of the other one.

tomaarsen replied:
You're right - I encountered the actual issue with the previous comment, and I extrapolated the fix to this one, not realising that there was an `[0]`.

warner-benjamin committed:
small bug fixes

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1251 @@
+        intermediate_size=1152,
+        num_hidden_layers=22,
+        num_attention_heads=12,
+        hidden_activation="gelu_python",
```
```suggestion
        hidden_activation="gelu",
```
`gelu` matches the research implementation. There's slight discrepancies otherwise with `gelu_python`.

warner-benjamin replied:
fixed with the latest push

warner-benjamin committed:
fixes

tomaarsen commented:
With all of these changes in place, I was able to confirm that the output to one of the trained models using the original research implementation nearly matches the output of the `transformers` ModernBERT-converted model. The only difference that remains is that the research implementation fuses the `self.mlp(self.mlp_norm(...))` using `@torch.compile(dynamic=True)`. I also only tested with a small input - I still have to test 1) larger inputs, 2) truncation, 3) batches, etc.

Do we allow something like this to get an exact 1-1 match?
```python
    @torch.compile(dynamic=True)
    def compiled_mlp(self, hidden_states: torch.Tensor) -> torch.Tensor:
        return self.mlp(self.mlp_norm(hidden_states))
```

- Tom Aarsen

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1231 @@
+from typing import Optional, Tuple, Union
+
+import torch
+import torch.nn as nn
```
```suggestion
from torch import nn
```
a nit, but the `as nn` is a bit unnecessary, let's avoid it.

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1231 @@
+        if reset_params:
+            self.attn_norm.reset_parameters()
+            self.mlp_norm.reset_parameters()
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+    ) -> torch.Tensor:
+        """Forward pass for a ModernBert layer, including both attention and MLP.
+
+        Args:
+            hidden_states: (total_nnz, dim)
+            position_ids: (total_nnz,)
+            attention_mask: (batch, max_seqlen)
+            cu_seqlens: (batch + 1,)
+            max_seqlen: int
+        """
+        attn_out = hidden_states + self.attn(
+            self.attn_norm(hidden_states),
+            position_ids=position_ids,
+            cu_seqlens=cu_seqlens,
+            max_seqlen=max_seqlen,
+            attention_mask=attention_mask,
+        )
+        return attn_out + self.mlp(self.mlp_norm(attn_out))
```
```suggestion

    @torch.compile(dynamic=True)
    def compiled_mlp(self, hidden_states: torch.Tensor) -> torch.Tensor:
        return self.mlp(self.mlp_norm(hidden_states))

    def forward(
        self,
        hidden_states: torch.Tensor,
        position_ids: Optional[torch.LongTensor] = None,
        attention_mask: Optional[torch.Tensor] = None,
        cu_seqlens: Optional[torch.Tensor] = None,
        max_seqlen: Optional[int] = None,
    ) -> torch.Tensor:
        """Forward pass for a ModernBert layer, including both attention and MLP.

        Args:
            hidden_states: (total_nnz, dim)
            position_ids: (total_nnz,)
            attention_mask: (batch, max_seqlen)
            cu_seqlens: (batch + 1,)
            max_seqlen: int
        """
        attn_out = hidden_states + self.attn(
            self.attn_norm(hidden_states),
            position_ids=position_ids,
            cu_seqlens=cu_seqlens,
            max_seqlen=max_seqlen,
            attention_mask=attention_mask,
        )
        return attn_out + self.compiled_mlp(attn_out)
```
Required to match research implementation's performance. With this in place, the difference between the original outputs and ModernBert become:
```
tensor([[[0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         ...,
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.]]], device='cuda:0',
       dtype=torch.bfloat16, grad_fn=<SubBackward0>)
Min: 0.0
Max: 0.0
Mean: 0.0
Std: 0.0
```

tomaarsen committed:
Update import

tomaarsen committed:
Use compiled mlp->mlp_norm to match research implementation

tomaarsen committed:
Propagate changes in modular to modeling

tomaarsen committed:
Replace duplicate attn_out_dropout in favor of attention_dropout

cc @warner-benjamin let me know if the two should remain separate!

tomaarsen committed:
Update BOS to CLS and EOS to SEP

Please confirm @warner-benjamin

tomaarsen committed:
Set default classifier bias to False, matching research repo

tomaarsen committed:
Update tie_word_embeddings description

tomaarsen committed:
Fix _init_weights for ForMaskedLM

NohTow commented on src/transformers/models/modernbert/configuration_modernbert.py:
```diff
@@ -0,0 +1,183 @@
+        classifier_pooling="mean",
+        classifier_norm=True,
+        classifier_bias=False,
+        classifier_activation=None,
```
I think the default should be "gelu"

NohTow commented on src/transformers/models/modernbert/configuration_modernbert.py:
```diff
@@ -0,0 +1,183 @@
+        unpad_no_grad=True,
+        decoder_bias=True,
+        classifier_dropout=0.0,
+        classifier_pooling="mean",
```
I think we ran most of the evaluations using CLS pooling (although I think we have better results with mean pooling)

tomaarsen replied:
Fixed in b3b4028e826d14b623bc6c35f3541ab51a67b234

tomaarsen committed:
Match base_model_prefix

tomaarsen committed:
Add compiled_head to match research repo outputs

tomaarsen committed:
Fix imports for ModernBertForMaskedLM

tomaarsen committed:
Just use "gelu" default outright for classifier

tomaarsen committed:
Fix config name typo: initalizer -> initializer

tomaarsen committed:
Remove some unused parameters in docstring. Still lots to edit there!

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1233 @@
+        embedding_dropout=0.0,
+        mlp_bias=False,
+        mlp_dropout=0.0,
+        unpad_inputs=True,
```
In my opinion, this shouldn't be an argument, but instead automatically set based on `config._attn_implementation`. E.g. `False` for "eager", `True` for "flash_attention_2".

tomaarsen replied:
Added this in 5e3d61d7180540003f5753c3f4186b456d160291

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1233 @@
+                else:
+                    input_ids, indices, cu_seqlens, max_seqlen, *_ = self._unpad_inputs(input_ids, attention_mask)
+        elif position_ids is None:
+            position_ids = torch.arange(seq_len, device=input_ids.device)
```
I *suspect* that this will have to be unsqueezed or batch-ified somewhere.

tomaarsen replied:
Added via 5e3d61d7180540003f5753c3f4186b456d160291

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1233 @@
+            position_ids = torch.arange(seq_len, device=input_ids.device)
+
+        hidden_states = self.embeddings(input_ids)
+
```
We don't have the eager attention yet, but it probably requires expanding the attention mask to 4D somewhere around here, e.g.:
```python
        # expand attention_mask
        if self.config._attn_implementation == "eager" and attention_mask is not None:
            # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
            attention_mask = _prepare_4d_attention_mask(attention_mask, hidden_states.dtype)
```

I didn't get the other attentions working yet @warner-benjamin, perhaps you can try and tackle those.

tomaarsen replied:
Added via a2051d6b4a0c771ecd36ba05603fe5823110acb5

tomaarsen committed:
Compile the embeddings forward

Not having this resulted in very slight differences - so small it wasn't even noticed for the base model, only for the large model.

But the tiny difference for large propagated at the embedding layer through the rest of the model, leading to notable differences of ~0.0084 average per value, up to 0.2343 for the worst case.

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1234 @@
+
+        loss = None
+        if labels is not None:
+            loss = self.loss_function(logits, labels, vocab_size=self.config.vocab_size)
```
`self.loss_function` does not exist, I believe.

ArthurZucker replied:
it does! 🤗 the loss_utils was correctly editted!

tomaarsen committed:
Add drafts for ForSequenceClassification/ForTokenClassification

tomaarsen committed:
Add initial SDPA support (not exactly equivalent to FA2 yet!)

During testing, FA2 and SDPA still differ by about 0.0098 per value in the token embeddings. It still predicts the correct mask fills, but I'd like to get it fully 1-1 if possible.

tomaarsen committed:
Only use attention dropout if training

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1450 @@
+    def __init__(
+        self,
+        config: ModernBertConfig,
+        pad_logits: bool = True,
+        pad_logits_no_grad: Optional[bool] = None,
+        sparse_prediction: bool = True,
+        sparse_pred_ignore_index: int = -100,
```
Let's try and get rid of these. Either via the `config`, or avoid needing them altogether.
I also feel like `pad_logits` should be inferrable via `config.unpad_inputs` and `pad_logits_no_grad` via `config.unpad_no_grad`.

I'm not sure what we should do with the `sparse_prediction` stuff.

tomaarsen replied:
Resolved via 38f959bf096bd05e1174ed4495daac004bf115fa, always open to suggestions

tomaarsen committed:
Add initial eager attention support (also not equivalent to FA2 yet!)

Frustratingly, I also can't get eager to be equivalent to FA2 (or sdpa), but it does get really close, i.e. avg ~0.010 difference per value.

Especially if I use fp32 for both FA2&eager, avg ~0.0029 difference per value

The fill-mask results are good with eager.

tomaarsen committed:
Add initial tests, output_attentions, output_hidden_states, prune_heads

Tests are based on BERT, not all tests pass yet: 23 failed, 79 passed, 100 skipped

tomaarsen committed:
Remove kwargs from ModernBertForMaskedLM

Disable sparse_prediction by default to match the normal HF, can be enabled via config

tomaarsen committed:
Remove/adjust/skip improper tests; warn if padding but no attn mask

tomaarsen committed:
Run formatting etc.

staghado commented:
I have a working first version with FlexAttention but for now there are still numerical differences between all three implementations.

tomaarsen commented:
@staghado does it give normal results with e.g. mask filling? If so, then feel free to push it into this branch (or if you don't have access, push it to a branch on your fork or something, and I can cherry-pick it into this PR).

Does it use unpadded or padded? I had a feeling like the SDPA/Eager attention differences compared to FA2 actually resulted when applying the sin/cos rotation, but for FlexAttention that would be equivalent as the FA2 path if you're using unpadded.

- Tom Aarsen

tomaarsen committed:
Run python utils/custom_init_isort.py

staghado commented:
The current FlexAttention PyTorch API requires batched query/key input so can't be used with unpadded inputs without pad/unpad logic.
I tried with both unpadding and padding and both exhibit differences with FA2.
I will push the code sometime tomorrow(still need to clean and do some more tests).
Mask filling performance across all three implementations is very similar but the probabilities are slightly off, here is a small table :

| impl          | prediction | prob      |
|----------------|-----------|------------|
| FA2            | tree      | 0.09228515625 |
| FA2            | trees     | 0.07177734375 |
| FA2            | ruins     | 0.04638671875 |
| FlexAttention  | tree      | 0.08349609375 |
| FlexAttention  | trees     | 0.0693359375 |
| FlexAttention  | ruins     | 0.047607421875 |
| SDPA           | tree      | 0.09033203125 |
| SDPA           | trees     | 0.0751953125 |
| SDPA           | ruins     | 0.04833984375 |

staghado committed:
FlexAttention with unpadded sequences(matches FA2 within bf16 numerics)

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+def _init_modernbert_weights(
+    config: ModernBertConfig,
+    module: Union[nn.Linear, nn.Embedding],
+    module_type: ModernBertModuleType,
+) -> None:
+    """
+    Initialize weights of a linear or embedding module.
+
+    :param config: The model config.
+    :param module: The linear or embedding submodule to initialize.
+    """
+    if module_type is None:
+        raise RuntimeError("When using the full megatron init, every module must have a type.")
+
+    cutoff_factor = config.initializer_cutoff_factor
+    if cutoff_factor is None:
+        cutoff_factor = 3
+
+    if module_type == ModernBertModuleType.in_module:
+        std = config.initializer_range  # for att_proj (same as QKV), ff_proj
+    elif module_type == ModernBertModuleType.out_module:
+        std = config.initializer_range / math.sqrt(2.0 * config.num_hidden_layers)  # for attn_out, ff_out
+    elif module_type == ModernBertModuleType.embedding:
+        std = config.initializer_range  # token embeddings (wte)
+    elif module_type == ModernBertModuleType.final_out:
+        std = config.hidden_size**-0.5  # final output (ff_out)
+    else:
+        raise RuntimeError(f"Unknown module type '{module_type}'")
+
+    nn.init.trunc_normal_(
+        module.weight,
+        mean=0.0,
+        std=std,
+        a=-cutoff_factor * std,
+        b=cutoff_factor * std,
+    )
+
+    if isinstance(module, nn.Linear):
+        if module.bias is not None:
+            nn.init.zeros_(module.bias)
```
all of this should go in the `def _init_weights(self, module):` function of the `ModerBertPreTrainedModel` and we should remove the ModuleTypes enumerations.

tomaarsen replied:
If we only have one `def _init_weights(self, module):` in `ModernBertPreTrainedModel`, how do we support the different desired `std`'s for the different kinds of linears/embeddings?

ArthurZucker replied:
you check based on the path of the module

tomaarsen replied:
Resolved in a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+            self.unpad_inputs = self._attn_implementation == "flash_attention_2"
+
+
+class ModernBertModuleType(str, Enum):
+    in_module = "in"
+    out_module = "out"
+    embedding = "emb"
+    final_out = "final_out"
+
+
+class ModernBertPoolingType(str, Enum):
+    cls = "cls"
+    mean = "mean"
+    max = "max"
```
to remove

tomaarsen replied:
Removed in a356708e9be7ad4da9d9c6e042e14fa5028eee36 and b3b4028e826d14b623bc6c35f3541ab51a67b234

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+            nn.init.zeros_(module.bias)
+
+
+def _unpad_modernbert_input(
```
correct me if I am wrong, the inputs are the same as any other LLMs no? In that case if you want to unpad you should be using https://github.com/huggingface/transformers/blob/1b6cb1eefc140ea2dbd74e23389050ebcb7f417d/src/transformers/modeling_flash_attention_utils.py#L63-L69 available in the utils!

tomaarsen replied:
I'm afraid not quite. We're unpadding much earlier and repadding much later, so that e.g. even the MaskedLM can take advantage of it. As a result, `_upad_input` (and `_flash_attention_forward`) aren't viable here.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+    return unpadded_inputs, indices, cu_seqlens, max_seqlen_in_batch, unpadded_position_ids, unpadded_labels
+
+
+def _pad_modernbert_output(
```
we actually also have utils for that here: https://github.com/huggingface/transformers/blob/1b6cb1eefc140ea2dbd74e23389050ebcb7f417d/src/transformers/modeling_flash_attention_utils.py#L310-L324

tomaarsen replied:
Ideally we'll avoid `flash_attn` their `pad_input` for the repadding, so that we can also support unpadded FlexAttention without the `flash_attn` dependency.
Currently `flash_attn` is still needed because of the `RotaryEmbedding` from `flash_attn`, but we're interested in a torch-only version as a fallback.

I will shrink `_pad_modernbert_output` though: we're not using the `padded_labels` anywhere.

tomaarsen replied:
Shrunk/removed unused path in b241a7e97a1413be7efcbf0e1435dd1581a2dc0f.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+    ):
+        # (total_nnz, 3, nheads, headdim)
+        total_nnz, three, nheads, headdim = qkv.shape
+        assert three == 3
```
no asserts in modeling code

tomaarsen replied:
Removed in 382e481dbb3df8ece138d81f2a6c8f91065210de

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        # (total_nnz, 3, nheads, headdim)
+        total_nnz, three, nheads, headdim = qkv.shape
+        assert three == 3
+        if qkv.is_contiguous():
```
let's maybe call qkv.contiguous() to just always have 1 path

tomaarsen replied:
Done in 5d05e8ebe6e9e54856440850769a89cc665acf8d

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+    return (attn.view(bs, dim),)
+
+
+# def flex_attention_forward(
+#     config: ModernBertConfig,
+#     query: torch.Tensor,
+#     key: torch.Tensor,
+#     value: torch.Tensor,
+#     mask: Optional[torch.Tensor],
+#     output_attentions: bool = False,
+#     **_kwargs,
+# ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+
+#     attn_output = flex_attention(
+#         query,
+#         key,
+#         value,
+#         enable_gqa=True,
+#         scale=config.scaling,
+#         return_lse=output_attentions,
+#     )
+#     if not output_attentions:
+#         attn_weights = None
+#     else:
+#         attn_output, attn_weights = attn_output
+
+#     attn_output = attn_output.transpose(1, 2).contiguous()
+#     return attn_output, attn_weights
+
```
to remove?

tomaarsen replied:
Added in f9301f4cf9886aa757c93f01d8c76e1421406722. Will check if there's utils for this in a bit.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        rope_theta = config.global_rope_theta
+        max_position_embeddings = config.max_position_embeddings
+        if self.local_attention != (-1, -1):
+            if config.local_rope_theta is not None:
+                rope_theta = config.local_rope_theta
+            max_position_embeddings = config.local_attention
+
```
can be set when you sel self.local_attention

ArthurZucker replied:
also rope should normally no be set in the attention layer but modeling layer, to just pass cos and sin.

tomaarsen replied:
> to just pass cos and sin

This is a bit tricky as the current ragged/unpadded rotary embedding implementation uses `flash_attn`'s `apply_rotary` which just applies the cos/sin rather than returning it for us to apply manually.

ArthurZucker replied:
Understood. In general I don't think the repo kernel brings that much more to the table than just having ragged inputs, which is the biggest win

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        self.out_drop = nn.Dropout(config.attention_dropout) if config.attention_dropout > 0.0 else nn.Identity()
+        self.pruned_heads = set()
+
+    def prune_heads(self, heads):
+        if len(heads) == 0:
+            return
+        heads, index = find_pruneable_heads_and_indices(heads, self.num_heads, self.head_dim, self.pruned_heads)
+
+        # Prune linear layers
+        self.Wqkv = prune_qkv_linear_layer(self.Wqkv, index)
+        self.Wo = prune_linear_layer(self.Wo, index, dim=1)
+
+        # Update hyper params and store pruned heads
+        self.num_heads = self.num_heads - len(heads)
+        self.all_head_size = self.head_dim * self.num_heads
+        self.pruned_heads = self.pruned_heads.union(heads)
+
+    def _init_weights(self, reset_params: bool = False):
+        _init_modernbert_weights(self.config, self.Wqkv, module_type=ModernBertModuleType.in_module)
+        _init_modernbert_weights(self.config, self.Wo, module_type=ModernBertModuleType.out_module)
+
```
to remove

tomaarsen replied:
Removed in 0a8d04458c7c8b110ae10d3500d7dc73cfc4bcb7 and a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+            bs = hidden_states.shape[0]
+            qkv = qkv.view(-1, 3, self.num_heads, self.head_dim)
+
+            attn_kwargs.update(
+                {
+                    "local_attention": self.local_attention,
+                    "cu_seqlens": cu_seqlens,
+                    "max_seqlen": max_seqlen,
+                    "bs": bs,
+                }
+            )
```
you shuld leave some of these as kwargs

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+            bs, seqlen = hidden_states.shape[:2]
+            qkv = qkv.view(bs, seqlen, 3, self.num_heads, self.head_dim)
+
+            attn_kwargs.update(
+                {
+                    "position_ids": position_ids,
+                    "attention_mask": attention_mask,
+                    "bs": bs,
+                    "seqlen": seqlen,
+                }
+            )
+
```
same here, the if else is useless as you can just pass kwargs and the function can ignore them if needed

tomaarsen replied:
Solved in 986c6feb9990b4e49c3ed810005e0855dd1c8b4f

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        )
+        hidden_states = attn_outputs[0]
+
+        return (self.out_drop(self.Wo(hidden_states)),) + attn_outputs[1:]  # add attentions if outputted
```
let's do the out projection before returning the output

tomaarsen replied:
Agreed, done in 5aaf273e5ea9eadff4b2b19efb4435874eabe270

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+    def __init__(self, config: ModernBertConfig, layer_id: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        if config.skip_first_prenorm and config.embedding_norm and layer_id == 0:
```
are there checkpoints released with `skip_first_prenorm` ? We need to remove all code paths that don't correspond to released checkpoints!

tomaarsen replied:
Actually, all release checkpoint will have `skip_first_prenorm` and `embedding_norm` as True, so I'll keep this one around.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        self.mlp_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.mlp = ModernBertMLP(config)
+
+    def _init_weights(self, reset_params: bool = False):
+        if reset_params:
+            self.attn_norm.reset_parameters()
+            self.mlp_norm.reset_parameters()
+
```
to remove

tomaarsen replied:
Removed in a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+            self.attn_norm.reset_parameters()
+            self.mlp_norm.reset_parameters()
+
+    @torch.compile(dynamic=True)
+    def compiled_mlp(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        return self.mlp(self.mlp_norm(hidden_states))
```
okay, I like this but I want to make sure: 
1. this does have an upside in perf? 
2. this works on `mps`, and any device!

tomaarsen replied:
I'd like for these `torch.compile` options to be optional (they were optional once too, we can reintroduce it), based on whether `torch.compile` can actually work (I believe it's not always possible on Windows for example).

warner-benjamin replied:
They need to be optional because torch.compile with dynamic shapes isn't compatible with DDP/FSDP/gradient checkpointing for all versions of PyTorch. Tom ported them from the research codebase because fusing the MLP norm into the MLP layer results in small numerical differences when in eager mode.

@ArthurZucker To answer your questions:
1. It's ~10% faster during pretraining
2. It's not [currently compatible with CPUs](https://github.com/triton-lang/triton-cpu), or MPS. AMD & Intel GPUs depend on the Triton version.

tomaarsen replied:
Made it optional in 37a60302048150d12fbf4cdc3691317c3a40ce21. It's a parameter defaulting to `None`, and if it stays `None`, then it's set based on whether `triton` is available. This seems to work very well on Windows, allowing me to use both no `triton` for no compilation and `triton` (via https://github.com/woct0rdho/triton-windows) for compilation.

ArthurZucker replied:
In general @tomaarsen it's the kind of thing we avoid, as we don't have a great way to do this today. We can merge this for now as we are in a rush, but we will be needing a refactoring later!

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        self.act = ACT2FN[config.classifier_activation] if config.classifier_activation else nn.Identity()
+        self.norm = (
+            nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+            if config.classifier_norm
```
same comment about codepathes!

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+            else nn.Identity()
+        )
+
+    def _init_weights(self, reset_params: bool = False):
+        if reset_params:
+            self.norm.reset_parameters()
+        _init_modernbert_weights(self.config, self.dense, module_type=ModernBertModuleType.in_module)
+
+    def reset_parameters(self):
+        self._init_weights(reset_params=True)
```
to remove

tomaarsen replied:
Removed in a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        self.act = ACT2FN[config.classifier_activation] if config.classifier_activation else nn.Identity()
+        self.norm = (
+            nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+            if config.classifier_norm
```
same comment

tomaarsen replied:
Removed in 5cd39ad940cdaa4875ca76f73cdb60b0c418ed4b

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        self.pooling_type = ModernBertPoolingType(config.classifier_pooling)
+
+    def forward(self, hidden_states: torch.Tensor, pool: Optional[bool] = True) -> torch.Tensor:
+        if pool:
+            if self.pooling_type == ModernBertPoolingType.cls:
+                output = hidden_states[:, 0]
+            elif self.pooling_type == ModernBertPoolingType.mean:
+                output = hidden_states.mean(dim=1)
+            elif self.pooling_type == ModernBertPoolingType.max:
+                output = hidden_states.max(dim=1)[0]
```
same here, we'd rather avoid this, even more in the forward pass. 
Either we have 3 different pooling heads, or you just define 3 functions like for attention! 
The best is to just have the 1 that is used in the released checkpoint

NohTow replied:
I think I'd vote for letting the possibility to choose, we evaluated with CLS pooling because it was the default (and standard across encoders), so the default will most likely be CLS, but mean pooling exhibited strong performance, so letting the choice is a nice feature imho.
(The released checkpoints are not CLS/mean pooling trained, only MLM).

I think 3 pooling heads is fine, but given the philosophy of FlexBert, maybe POOL2FN is more appropriate.

tomaarsen replied:
Separated into 2 defined functions, much like for Attention, in b3b4028e826d14b623bc6c35f3541ab51a67b234. We're keeping MEAN and CLS as they're both common: CLS is the usual default in HF, but MEAN works a smidge better for some of our tests.
The default is CLS for now.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+
+        return self.drop(self.norm(self.act(self.dense(output))))
+
+    def _init_weights(self, reset_params: bool = False):
+        _init_modernbert_weights(self.config, self.dense, module_type=ModernBertModuleType.out_module)
+        if reset_params and hasattr(self.norm, "reset_parameters"):
+            self.norm.reset_parameters()
+
+    def reset_parameters(self):
+        self._init_weights(reset_params=True)
```
to remove as well

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        for layer, heads in heads_to_prune.items():
+            self.layers[layer].attn.prune_heads(heads)
+
+    def _init_weights(self, module: Optional[nn.Module] = None, reset_params: Optional[bool] = None):
+        if module and hasattr(module, "_init_weights"):
+            super()._init_weights(module, reset_params=reset_params)
+        elif isinstance(reset_params, bool):
+            self.embeddings._init_weights(reset_params=reset_params)
+
+            if reset_params:
+                self.final_norm.reset_parameters()
```
everything should be handled in the super def of this function!

tomaarsen replied:
Removed in a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        if self.config.unpad_inputs:
+            if indices is None and cu_seqlens is None and max_seqlen is None:
+                repad = True
+                if self.config.unpad_no_grad:
```
not sure I understand why you would wnat gradients when padding / unpadding?

warner-benjamin replied:
You'd want gradients for unpadding the logits in the MLM head if you want to calculate your own loss outside of the transformers model.

ArthurZucker replied:
I might be missing something but this does not make sense to me, padding / unpadding should be transparent with regards to the loss whatsoever. It's a way of feeding the inputs to the models, completely independent of the weights of the model

ArthurZucker replied:
This needs to be removed IMO!

tomaarsen replied:
Removed in 44275fd738741895362f6bd8fe5ade45f5c6f98d

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        self.model = ModernBertModel(config)
+        self.head = ModernBertPredictionHead(config)
+
+        if config.tie_word_embeddings:
+            decoder_weights = self.model.embeddings.tok_embeddings.weight
```
the way to do this is by defining the `    _tied_weights_keys = ["model.embeddings.tok_embeddings.weight"]`

tomaarsen replied:
Fixed in b444c15e3936036a7a1100b44fedb61f0c923261

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        # Initialize weights and apply final processing
+        self._init_weights(reset_params=False)
+
+    def _init_weights(self, module: Optional[nn.Module] = None, reset_params: Optional[bool] = None):
+        assert (module is None) != (reset_params is None), "arg module xor reset_params must be specified"
+        if module:
+            super()._init_weights(module, reset_params=reset_params)
+        else:
+            assert isinstance(reset_params, bool)
+            self.model._init_weights(reset_params=reset_params)
+            self.head._init_weights(reset_params=reset_params)
+
+            # Output weights.
+            if not self.config.tie_word_embeddings:
+                _init_modernbert_weights(self.config, self.decoder, module_type=ModernBertModuleType.out_module)
```
to remove

tomaarsen replied:
Removed in a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+    ) -> Union[Tuple[torch.Tensor], MaskedLMOutput]:
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if self.config.unpad_inputs:
+            if indices is None and cu_seqlens is None and max_seqlen is None:
+                batch_size, seq_len = input_ids.shape[:2]
+                if self.config.unpad_no_grad:
+                    input_ids, indices, cu_seqlens, max_seqlen, position_ids, labels = self._unpad_inputs_no_grad(
+                        input_ids=input_ids, attention_mask=attention_mask, position_ids=position_ids, labels=labels
+                    )
+                else:
+                    input_ids, indices, cu_seqlens, max_seqlen, position_ids, labels = self._unpad_inputs(
+                        input_ids=input_ids, attention_mask=attention_mask, position_ids=position_ids, labels=labels
+                    )
```
it happens in the modeling code eitherway so this is repeated

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        if labels is not None:
+            loss = self.loss_function(logits, labels, vocab_size=self.config.vocab_size)
+
+        if self.config.unpad_inputs:
+            if self.config.unpad_no_grad:
+                logits, _ = self._pad_outputs_no_grad(logits, indices, batch_size, seq_len)
+            else:
+                logits, _ = self._pad_outputs(logits, indices, batch_size, seq_len)
```
I don't think it should be config dependent , but rather `self.padded_inputs` or something. If we want to leverage by default the strength of unpadded kernels and etc.

tomaarsen replied:
Since fb748ce728af891fbf56ea77fcea4f3aa2443ff3 it's just based on whatever attention implementation is used. It doesn't make sense to unpad with SDPA/Eager and it doesn't make sense to pad with FA2/Flex (in our implementation), so I removed `unpad_inputs` altogether.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        # Initialize weights and apply final processing
+        self._init_weights(reset_params=False)
+
+    def _init_weights(self, module: Optional[nn.Module] = None, reset_params: Optional[bool] = None):
+        assert (module is None) != (reset_params is None), "arg module xor reset_params must be specified"
+        if module:
+            super()._init_weights(module, reset_params=reset_params)
+        else:
+            assert isinstance(reset_params, bool)
+            self.model._init_weights(reset_params=reset_params)
+            self.head._init_weights(reset_params=reset_params)
+            _init_modernbert_weights(self.config, self.classifier, module_type=ModernBertModuleType.final_out)
```
same commetn

tomaarsen replied:
Removed in a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+        # Initialize weights and apply final processing
+        self._init_weights(reset_params=False)
+
+    def _init_weights(self, module: Optional[nn.Module] = None, reset_params: Optional[bool] = None):
+        assert (module is None) != (reset_params is None), "arg module xor reset_params must be specified"
+        if module:
+            super()._init_weights(module, reset_params=reset_params)
+        else:
+            assert isinstance(reset_params, bool)
+            self.model._init_weights(reset_params=reset_params)
+            _init_modernbert_weights(self.config, self.classifier, module_type=ModernBertModuleType.final_out)
```
same!

tomaarsen replied:
Removed in a356708e9be7ad4da9d9c6e042e14fa5028eee36

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+
+
+def eager_attention_forward(
+    self: "ModernBertAttention",
```
```suggestion
    module: "ModernBertAttention",
```

tomaarsen replied:
Tackled in f83fdc0cb09c72e44182d8a03761402f96ad3b4d

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1515 @@
+    # if is_flash_attn_2_available():
+
+
+class ApplyRotaryEmbUnpad(torch.autograd.Function):
```
TBH, regarding applying rotary to ragged inputs, the formulation we wrote in `LlamaRotaryEmbedding` still holds. 
Just adding an extra dummy dimension [1, ragged_length, ...] works! 
We are working on this in #35235

staghado replied:
but it's slower/more memory if we use the PyTorch one no?

ArthurZucker reviewed (commented):
Good work! 
Is the prunning necessary or is it to follow previous implementations? 🤗

tomaarsen commented:
> Is the prunning necessary or is it to follow previous implementations? 🤗

Not necessary at all. I just added it to 1) match BERT a bit better and 2) allow the BERT tests to pass. I'm totally fine with removing it fully.

tomaarsen committed:
Reformat init_weights based on review

tomaarsen committed:
self -> module in attention forwards

tomaarsen committed:
Remove if config.tie_word_embeddings

tomaarsen committed:
Reformat output projection on a different line

tomaarsen committed:
Remove pruning

tomaarsen committed:
Remove assert

tomaarsen committed:
Call contiguous() to simplify paths

tomaarsen committed:
Remove prune_qkv_linear_layer

tomaarsen committed:
Format code

staghado commented on src/transformers/models/modernbert/modeling_modernbert.py:
```diff
@@ -67,45 +67,22 @@ def forward(
-                interleaved=False,
-                inplace=True,
-            )
+        qkv = qkv.contiguous()
```
calling contiguous is not free, no?

tomaarsen replied:
I don't think it is, no. I'd rather keep the faster solution, but I understand @ArthurZucker's desire for simplicity.

staghado replied:
without additional tests that the performance does not degrade, I would keep the contiguity check.

NohTow replied:
I am not 100% sure and I don't think a test is that hurtful, but it seems like `.contiguous()` is a [no-op if the tensor is already contiguous](https://discuss.pytorch.org/t/does-contiguous-creates-a-copy-of-the-memory/67656/2).
The answer is from 2020, but the [PyTorch documentation seems to agree](https://pytorch.org/docs/stable/generated/torch.Tensor.contiguous.html) and a [check seems to be performed](https://github.com/pytorch/pytorch/blob/ccf35af1421447c6425d6d290cb001b80d93c99b/torch/_refs/__init__.py#L2992).

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1423 @@
+            init_weight(module.Wqkv, stds["in"])
+            init_weight(module.Wo, stds["out"])
+        elif isinstance(module, ModernBertPredictionHead):
+            init_weight(module.dense, stds["in"])  # TODO: Should this be "out"/"final_out"?
```
@warner-benjamin "in" is taken from the research implementation, but it's a bit odd to me. Could you double-check this and confirm that "in" is correct?

tomaarsen commented:
@ArthurZucker due to small differences between FA2/SDPA/Flex/Eager, we would love to default to FA2 if `flash_attn` is installed. Would this be possible via some config option, or do we *have* to rely on users to manually specify `attn_implementation="flash_attention_2"`?

tomaarsen committed:
Keep as kwargs, only use if needed

tomaarsen committed:
Remove unused codepaths & related config options

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1369 @@
+    qkv: torch.Tensor,
+    rotary_emb: ModernBertUnpaddedRotaryEmbedding,
+    cu_seqlens: torch.Tensor,
+    block_mask: BlockMask,
```
```suggestion
    block_mask: "BlockMask",
```
Ensure that this doesn't crash if torch < 2.5 and `BlockMask` doesn't exist. Same in another method below.
cc @NohTow

tomaarsen replied:
Resolved in 3a24af438d0c4ae6bc8a4443a93f72f5428144c0

tomaarsen committed:
Remove 3d attn_mask test; fix token classification tuple output

tomaarsen committed:
Reorder: attention_mask above position_ids, fixes gradient checkpointing

tomaarsen committed:
Merge branch 'main' into pr-35158

tomaarsen committed:
Fix usage if no FA2 or torch v2.5+

tomaarsen commented:
These are the remaining test failures:
```
FAILED tests/models/modernbert/test_modeling_modernbert.py::ModernBertModelTest::test_cpu_offload - torch._dynamo.exc.InternalTorchDynamoError: AttributeError: 'dict' object has no attribute 'node'
FAILED tests/models/modernbert/test_modeling_modernbert.py::ModernBertModelTest::test_disk_offload_bin - torch._dynamo.exc.InternalTorchDynamoError: AttributeError: 'dict' object has no attribute 'node'
FAILED tests/models/modernbert/test_modeling_modernbert.py::ModernBertModelTest::test_disk_offload_safetensors - torch._dynamo.exc.InternalTorchDynamoError: AttributeError: 'dict' object has no attribute 'node'
FAILED tests/models/modernbert/test_modeling_modernbert.py::ModernBertModelTest::test_resize_tokens_embeddings - AssertionError: expected size 84==109, stride 32==32 at dim=0
FAILED tests/models/modernbert/test_modeling_modernbert.py::ModernBertModelTest::test_retain_grad_hidden_states_attentions - RuntimeError: This backward function was compiled with non-empty donated buffers which requires create_graph=False and retain_graph=False. Please keep backwar...
```
with these reasons:

* `test_cpu_offload`, `test_disk_offload_bin`, `test_disk_offload_safetensors`: Unsure, something about 1) torch.compile and 2) some weight being on the meta device, but without a value to put it back on CUDA.
* `test_resize_tokens_embeddings`: Caused by torch.compile breaking with resizing
* `test_retain_grad_hidden_states_attentions`: Caused by torch.compile, though this might be fixable with some parameters to the compile call.

They don't strictly all have to be fixed, but we can also disable those tests and give the users warnings when they do something that doesn't work with this architecture (e.g. resizing tokens seems to not work currently?).

cc @warner-benjamin

tomaarsen committed:
Make torch.compile/triton optional

Should we rename 'compile'? It's a bit vague

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1390 @@
+        self.deterministic_flash_attn = deterministic_flash_attn
+        self.sparse_prediction = sparse_prediction
+        self.sparse_pred_ignore_index = sparse_pred_ignore_index
+        self.compile = compile
+
+        if self.compile is None:
+            self.compile = is_triton_available()
```
I'm open to suggestions for:
* the arg name, `compile` feels a bit vague, but it's not exactly `compile_layers` either.
* when to set `compile` to True/False. `is_triton_available` is a good start for Windows, but I'm not sure if it'll work well with MPS/FSDP/DDP out of the box.

cc @warner-benjamin

warner-benjamin replied:
call it something along the lines of `reference_compiling` since we only do it to match the reference implementation?

tomaarsen commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1390 @@
+        counts = offsets[1:] - offsets[:-1]
+        return torch.repeat_interleave(torch.arange(len(counts), device=device, dtype=torch.int32), counts)
+
+    @torch.compile(dynamic=False)
```
@staghado do we need this compile? We now have a `compile` argument that we need to abide by. Ideally I'd remove this compile unless it's necessary for matching the research performance or if it helps speed a lot.

Otherwise we can't get Flex Attention on e.g. Windows or MPS (currently we can't either way, due to the FA2 dependency for Flex).

staghado replied:
we can set it according to the compile argument or remove it for now. We can add it later when it's more stable

tomaarsen replied:
I've blanket removed it here: 1e367df9b9e0a92189c562445d42d09b859c22d1
In the other cases with `torch.compile`, we have to duplicate all the code in the function, which was a bit much here.

tomaarsen committed:
Separate pooling options into separate functions (cls, mean) - cls as default

tomaarsen committed:
Simplify _pad_modernbert_output, remove unused labels path

tomaarsen committed:
Update tied weights to remove decoder.weight, simplify decoder loading

tomaarsen committed:
Adaptively set config.compile based on hf_device_map/device/resize, etc.

tomaarsen committed:
Merge branch 'main' of https://github.com/huggingface/transformers into pr-35158

tomaarsen committed:
Update ModernBertConfig docstring

tomaarsen committed:
Satisfy some consistency checks, add unfinished docs

tomaarsen committed:
Merge branch 'main' of https://github.com/huggingface/transformers into pr-35158

tomaarsen committed:
Only set compile to False if there's more than 1 device

tomaarsen committed:
Add docstrings for public ModernBert classes

tomaarsen committed:
Dont replace docstring returns - ends up being duplicate

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35158). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

tomaarsen committed:
Fix mistake in toctree

tomaarsen committed:
Reformat toctree

tomaarsen committed:
Patched FlexAttention, SDPA, Eager with Local Attention

tomaarsen committed:
Implement FA2 -> SDPA -> Eager attn_impl defaulting, crucial

both to match the original performance, and to get the highest inference speed without requiring users to manually pick FA2

tomaarsen committed:
Patch test edge case with Idefics3 not working with 'attn_implementation="sdpa"'

tomaarsen committed:
Repad all_hidden_states as well

warner-benjamin committed:
rename config.compile to reference_compile

warner-benjamin committed:
disable flex_attention since it crashes

bclavie committed:
Update modernbert.md

NohTow committed:
Using dtype min to mask in eager

tomaarsen committed:
Fully remove flex attention for now

It's only compatible with the nightly torch 2.6, so we'll leave it be for now. It's also slower than eager/sdpa.

Also, update compile -> reference_compile in one more case

tomaarsen committed:
Call contiguous to allow for .view()

ArthurZucker commented on docs/source/en/model_doc/modernbert.md:
```diff
@@ -0,0 +1,96 @@
+<!--Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```

ArthurZucker commented on src/transformers/models/modernbert/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+_import_structure = {
+    "configuration_modernbert": ["ModernBertConfig"],
+}
+
+try:
+    if not is_torch_available():
+        raise OptionalDependencyNotAvailable()
+except OptionalDependencyNotAvailable:
+    pass
+else:
+    _import_structure["modeling_modernbert"] = [
+        "ModernBertForMaskedLM",
+        "ModernBertModel",
+        "ModernBertPreTrainedModel",
+        "ModernBertForSequenceClassification",
+        "ModernBertForTokenClassification",
+    ]
+
+if TYPE_CHECKING:
+    from .configuration_modernbert import ModernBertConfig
+
+    try:
+        if not is_torch_available():
+            raise OptionalDependencyNotAvailable()
+    except OptionalDependencyNotAvailable:
+        pass
+    else:
+        from .modeling_modernbert import (
+            ModernBertForMaskedLM,
+            ModernBertForSequenceClassification,
+            ModernBertForTokenClassification,
+            ModernBertModel,
+            ModernBertPreTrainedModel,
+        )
+
+else:
+    import sys
+
+    sys.modules[__name__] = _LazyModule(__name__, globals()["__file__"], _import_structure, module_spec=__spec__)
```
```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from typing import TYPE_CHECKING

from ...utils import _LazyModule
from ...utils.import_utils import define_import_structure


if TYPE_CHECKING:
    from .configuration_modernbert import *
    from .modeling_modernbert  import *
    from .processing_modernbert  import *
else:
    import sys

    _file = globals()["__file__"]
    sys.modules[__name__] = _LazyModule(__name__, _file, define_import_structure(_file), module_spec=__spec__)

```

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+        self.config = config
+        self.Wi = nn.Linear(config.hidden_size, int(config.intermediate_size) * 2, bias=config.mlp_bias)
+        self.act = ACT2FN[config.hidden_activation]
+        self.drop = nn.Dropout(config.mlp_dropout) if config.mlp_dropout > 0.0 else nn.Identity()
```
we should just leave it, if dropout is == 0 it will be no op anyways no?

tomaarsen replied:
Fixed in aa1bdb4b951e785c35b1c2c27b11c850fc62d298

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+    scale = module.head_dim**-0.5
+    attn_weights = torch.matmul(query, key.transpose(2, 3)) * scale
+
+    if attention_mask is not None:
+        expanded_mask = _prepare_4d_attention_mask(attention_mask, attn_weights.dtype, tgt_len=key.shape[-2])
```
the attention mask should not be prepared in the eager pass, it should be created in the `Model` otherwise we are wasting a lot of compute

tomaarsen replied:
You're very right - I'm computing it on the Model level now since 4145119f29c541a33e79db37eec4689428742c1e. We also immediately compute the attention mask with sliding window there, and pass it along down to this attention function.
The attention results are identical, but inference is much quicker! Thanks.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+    if attention_mask is not None:
+        expanded_mask = _prepare_4d_attention_mask(attention_mask, attn_weights.dtype, tgt_len=key.shape[-2])
+
+        if local_attention != (-1, -1):
+            # Create position indices
+            rows = torch.arange(expanded_mask.shape[2]).unsqueeze(0)
+            # Calculate distance between positions
+            distance = torch.abs(rows - rows.T)
+
+            # Create sliding window mask (1 for positions within window, 0 outside)
+            window_mask = (distance <= local_attention[0]).unsqueeze(0).unsqueeze(0).to(attention_mask.device)
+            # Combine with existing mask
+            expanded_mask.masked_fill_(window_mask.logical_not(), torch.finfo(attn_weights.dtype).min)
```
same for sliding window, this should be handled the same way we do with mistral!

ArthurZucker replied:
it should also be standardized!

tomaarsen replied:
You're very right - I'm computing it on the Model level now since 4145119f29c541a33e79db37eec4689428742c1e. We also immediately compute the attention mask with sliding window there, and pass it along down to this attention function.

Both Eager and SDPA use the same attention mask: `float`-based with `0` meaning "include" and `torch.finfo(dtype).min` meaning "exclude".

The attention results are identical, but inference is much quicker! Thanks.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+
+    # upcast attention to fp32
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+    if module.training:
+        attn_weights = nn.functional.dropout(attn_weights, p=module.attention_dropout)
```
```suggestion
    attn_weights = nn.functional.dropout(attn_weights, p=module.attention_dropout, training=self.training))
```

tomaarsen replied:
Applied this fix in e5dca63e084242006667665456b0d6162f0ee9ab

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+    # query, key, value: [batch_size, heads, seq_len, head_dim]
+    query, key = apply_rotary_pos_emb(query, key, cos, sin)
+
+    if attention_mask is not None:
+        attention_mask = attention_mask[:, None, None, :].expand(
+            attention_mask.shape[0], 1, attention_mask.shape[1], attention_mask.shape[1]
+        )
+
+        if local_attention != (-1, -1):
+            # Create position indices
+            rows = torch.arange(attention_mask.shape[2]).unsqueeze(0)
+            # Calculate distance between positions
+            distance = torch.abs(rows - rows.T)
+
+            # Create sliding window mask (1 for positions within window, 0 outside)
+            window_mask = (distance <= local_attention[0]).unsqueeze(0).unsqueeze(0).to(attention_mask.device)
+            # Combine with existing mask
+            attention_mask = torch.logical_and(attention_mask, window_mask)
+
+        attention_mask = attention_mask.to(torch.bool)
```
same comment about the attention mask. When the layer is sliding we need to just fix the mask at the decoder layer call level see for gemma2:  https://github.com/huggingface/transformers/blob/deac971c469bcbb182c2e52da0b82fb3bf54cccf/src/transformers/models/gemma2/modeling_gemma2.py#L460-L472

tomaarsen replied:
I've moved it to the modeling layer instead in 4145119f29c541a33e79db37eec4689428742c1e to avoid having to compute it for each layer. We compute both the global attention mask and the attention mask with sliding window there, and pass it along down to the SDPA/Eager attention functions.

The attention results are identical, but inference is much quicker! Thanks.

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+        labels: Optional[torch.Tensor] = None,
+    ):
+        return self._unpad_inputs(
+            input_ids=input_ids, attention_mask=attention_mask, position_ids=position_ids, labels=labels
+        )
+
+    def _unpad_inputs(
+        self,
+        input_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        position_ids: Optional[torch.Tensor] = None,
+        labels: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor | int | None]:
+        return _unpad_modernbert_input(
+            inputs=input_ids, attention_mask=attention_mask, position_ids=position_ids, labels=labels
+        )
+
+    @torch.no_grad()
+    def _pad_outputs_no_grad(
+        self,
+        inputs: torch.Tensor,
+        indices: torch.Tensor,
+        batch_size: int,
+        seqlen: int,
+    ):
+        return self._pad_outputs(
+            inputs=inputs,
+            indices=indices,
+            batch_size=batch_size,
+            seqlen=seqlen,
+        )
+
+    def _pad_outputs(
+        self,
+        inputs: torch.Tensor,
+        indices: torch.Tensor,
+        batch_size: int,
+        seqlen: int,
+    ):
+        return _pad_modernbert_output(inputs=inputs, indices=indices, batch=batch_size, seqlen=seqlen)
```
in strong favor of not having these 😉 one liners to call a function that already exists

warner-benjamin replied:
@staghado removed them

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+            init_weight(module.classifier, stds["final_out"])
+
+    @classmethod
+    def _autoset_attn_implementation(
```
not sure this one should be re implemented!

warner-benjamin replied:
We want ModernBERT to default to FA2 if possible because it's the only path which supports full model unpadding, which is a significant portion of our speedup on variable sequence length batches. 

SPDA and Eager won't support full model unpadding, just partial, so they will always be slower and use more memory. FlexAttention should support full model unpadding, but we are delaying that path for a future PR.

(The FA path also matches our research/training codebase and outputs)

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -0,0 +1,1608 @@
+    def forward(
+        self, hidden_states: torch.Tensor, attention_mask: torch.Tensor, pool: Optional[bool] = True
+    ) -> torch.Tensor:
+        if pool:
```
is pool ever set to `False` somewhere?

NohTow replied:
I don't think so.
To me we only use it for sequence classification and I am not really sure why we would use this head without pooling.

tomaarsen replied:
I'll fix this, one minute

tomaarsen replied:
Fixed in 0e572d503f95309a9c6ebb2fd95a84d27722a9b6

ArthurZucker reviewed (commented):
For all the torch.compile calls that are here, we'd rather have users use torch.compile model and reap the benefits without impacting other users! I believe it should bring also more to the table in terms of optimisation! 

Cool to use the rope kernels, would be nice to have data on how much better it is, but let's also note that it might have issues with bfloat16. We have a version that forces computation of sin and cos because cached ones led to errors! 🤗

tomaarsen committed:
Copyright 2020 -> 2024

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

tomaarsen committed:
Update/simplify __init__ structure

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

tomaarsen committed:
Remove "... if dropout_prob > 0 else identity"

As dropout with 0.0 should be efficient like identity

staghado commented:
the model with FA2 and the RoPE kernel are not torch.compile compatible, we can't compile the whole model while using these.

staghado committed:
re-use existing pad/unpad functions instead of creating new ones

staghado committed:
remove flexattention method

warner-benjamin commented:
> the model with FA2 and the RoPE kernel are not torch.compile compatible, we can't compile the whole model while using these.

FA2 is compatible now, but the FA RoPE kernel isn't yet. I have a in progress fix I need to get merged into the FA repo.

tomaarsen committed:
Compute attention_mask and local_attention_mask once in modeling

tomaarsen committed:
Simplify sequence classification prediction heads, only CLS now

Users can make custom heads if they feel like it

Also removes the unnecessary pool parameter

tomaarsen committed:
Simplify module.training in eager attn

tomaarsen committed:
Also export ModernBertPreTrainedModel

tomaarsen committed:
Update the documentation with links to finetuning scripts

tomaarsen committed:
Explain local_attention_mask parameter in docstring

tomaarsen committed:
Simplify _autoset_attn_implementation, rely on super()

tomaarsen committed:
Keep "in" to initialize Prediction head

Doublechecked with Benjamin that it's correct/what we used for pretraining

warner-benjamin committed:
add back mean pooling

warner-benjamin committed:
Use the pooling head in TokenClassification

warner-benjamin commented:
We want mean pooling as an option for classification because Local Attention means unlike BERT the CLS token doesn't see all the output in all the attention layers, so mean pooling could outperform CLS on greater than 128 token sequences.

Also, I added the pooling head to TokenClassification because otherwise we are throwing away one pretrained linear layer `ModernBertPoolingHead.dense`.

warner-benjamin committed:
update copyright

tomaarsen committed:
Reset config._attn_implementation_internal on failure

warner-benjamin committed:
Allow optional attention_mask in ForMaskedLM head

warner-benjamin committed:
fix failing run_slow tests

tomaarsen committed:
Add links to the paper

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -799,10 +756,11 @@ def forward(
         attn_outputs = self.attn(
             self.attn_norm(hidden_states),
+            attention_mask=attention_mask,
+            local_attention_mask=local_attention_mask,
```
can we just switch to `sliding_window_mask` 😉

tomaarsen replied:
Renamed in d799d65bd9ebeab85e1d6a7f453841c5aa21b272

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -1428,6 +1280,39 @@ def forward(
         )
 
 
+def cls_pooling(hidden_states: torch.Tensor, attention_mask: torch.Tensor) -> torch.Tensor:
+    return hidden_states[:, 0]
+
+
+def mean_pooling(hidden_states: torch.Tensor, attention_mask: torch.Tensor) -> torch.Tensor:
+    return (hidden_states * attention_mask.unsqueeze(-1)).sum(dim=1) / attention_mask.sum(dim=1, keepdim=True)
+
+
+MODERNBERT_POOLING_FUNCTION = {
+    "cls": cls_pooling,
+    "mean": mean_pooling,
+}
+
```
as per our previous offline discussino let's only add one Head, with cLS if it's the one used, and no one liner for the pooling!

tomaarsen replied:
Tackled as discussed offline in 92e17c667aae8e9473237e1ec9a843f16eb58e70

ArthurZucker reviewed (approved):
Super cool! 2 things left: 1. remove the gradient thing: padding and unpadding is not model weight dependant, and should never have gradients. 
2. remove the 2 functions, and just call the Head ClsPooling or MeanPooling depending on the one that was release / most common cf our offline discussion @tomaarsen

tomaarsen committed:
Remove unpad_no_grad, always pad/unpad without gradients

tomaarsen committed:
local_attention_mask -> sliding_window_mask

tomaarsen committed:
Revert "Use the pooling head in TokenClassification"

This reverts commit 99c38badd1dbce01d7aef41095fbf2f5cce87279.

There was no real motivation, no info on whether having this bigger head does anything useful.

tomaarsen committed:
Simplify pooling, 2 options via if-else

ArthurZucker commented:
Great addition! Thanks all for your hard work! 🤗
